README.md
[![npm](https://img.shields.io/npm/v/@flitz/body.svg)](https://www.npmjs.com/package/@flitz/body) [![supported flitz version](https://img.shields.io/static/v1?label=flitz&message=0.14.0%2B&color=blue)](https://github.com/flitz-js/flitz) [![last build](https://img.shields.io/github/workflow/status/flitz-js/body/Publish)](https://github.com/flitz-js/body/actions?query=workflow%3APublish) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](https://github.com/flitz-js/body/pulls)

# @flitz/body

> A body parser and handler middleware for [flitz](https://github.com/flitz-js/flitz).

## Install

Run

```bash
npm install --save @flitz/body
```

from the folder, where your `package.json` is stored.

## Usage

```javascript
const flitz = require('flitz');
const body = require('@flitz/body');

const run = async () => {
  const app = flitz();

  //             👇👇👇
  app.post('/', [ body() ], async (req, res) => {
    res.write('Your body as string from buffer: ' + req.body.toString());
    res.end();
  });

  await app.listen(3000);
};

run();
```

Or the TypeScript way:

```typescript
import flitz from 'flitz';
import { body } from '@flitz/body';

const run = async () => {
  const app = flitz();

  //             👇👇👇
  app.post('/', [ body() ], async (req, res) => {
    res.write('Your body as string from buffer: ' + req.body.toString());
    res.end();
  });

  await app.listen(3000);
};

run();
```

### JSON

```typescript
import flitz from 'flitz';
import { json } from '@flitz/body';

const run = async () => {
  const app = flitz();

  //             👇👇👇
  app.post('/', [ json() ], async (req, res) => {
    res.write('Your JSON body as JavaScript object: ' + JSON.stringify(req.body, null, 2));
    res.end();
  });

  await app.listen(3000);
};

run();
```

### YAML

```typescript
import flitz from 'flitz';
import { yaml } from '@flitz/body';

const run = async () => {
  const app = flitz();

  //             👇👇👇
  app.post('/', [ yaml() ], async (req, res) => {
    res.write('Your YAML body as JavaScript object: ' + JSON.stringify(req.body, null, 2));
    res.end();
  });

  await app.listen(3000);
};

run();
```

### Form

```typescript
import flitz from 'flitz';
import { form } from '@flitz/body';

const run = async () => {
  const app = flitz();

  //             👇👇👇
  app.post('/', [ form() ], async (req, res) => {
    res.write('Your body as key / value pair object: ' + JSON.stringify(req.body, null, 2));
    res.end();
  });

  await app.listen(3000);
};

run();
```

### String

```typescript
import flitz from 'flitz';
import { string } from '@flitz/body';

const run = async () => {
  const app = flitz();

  //              👇👇👇
  app.post('/', [ string() ], async (req, res) => {
    res.write('Your body as UTF-8 string: ' + req.body);
    res.end();
  });

  await app.listen(3000);
};

run();
```

## Maximum size

```typescript
//                    👇👇👇 (128 MB)
app.post('/', [ body({ max: 128 * 1024 * 1024 }) ], async (req, res) => {
  // your code here
});
```

## TypeScript

TypeScript is optionally supported. The module contains its own [definition files](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html).

## Credits

The module makes use of:

* [JS-YAML](https://github.com/nodeca/js-yaml) by [Nodeca](https://github.com/nodeca)

## License

MIT © [Marcel Kloubert](https://github.com/mkloubert)


CHANGELOG.md
# @flitz/body

## 3.0.0

* [BREAKING CHANGE]: `form()` parses input binary as ASCII now

## 2.0.0

* add [YAML](https://www.npmjs.com/package/js-yaml) support
* [BREAKING CHANGE]: when parsing wrong JSON or YAML, 400 is returned by default
* can handle parse errors now
* `options` of parsers can also be `(null)` now

## 1.1.0

* cleanup package

## 1.0.1

* first stable release

## 0.6.2

* [BREAKING CHANGE]: `maxLength` => `max`

## 0.5.1

* plugin requires at least [flitz 0.8.0+](https://github.com/flitz-js/flitz) now
* can define `maxLength` property for parsers now
* add `form()` middleware
* `body` property is of type `any` now

## 0.4.0

* update donation links to [open collective](https://opencollective.com/flitz) and [PayPal](https://paypal.me/MarcelKloubert)

## 0.3.1

* update to [flitz 0.4.1](https://github.com/flitz-js/flitz)

## 0.2.5

* add `string()` parser

## 0.1.1

initial release


src/__tests__/body.ts
import request from 'supertest';
import { body } from '..';

it('should parse request body to a Buffer', async () => {
  const app = global.createTestApp();

  app.post('/', body(), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('is büffer: ' + Buffer.isBuffer(req.body) + ' ' + req.body.length);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send('0123456789');

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('is büffer: true 10');
});

it('should parse request body to a Buffer and return 200 if max = 10 and body = 9', async () => {
  const app = global.createTestApp();

  app.post('/', body({ max: 10 }), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('is büffer: ' + Buffer.isBuffer(req.body) + ' ' + req.body.length);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send('123456789');

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('is büffer: true 9');
});

it('should parse request body to a Buffer and return 200 if max = 10 and body = 10', async () => {
  const app = global.createTestApp();

  app.post('/', body({ max: 10 }), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('is büffer: ' + Buffer.isBuffer(req.body) + ' ' + req.body.length);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send('0123456789');

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('is büffer: true 10');
});

it('should parse request body to a Buffer and return 413 if body is bigger than max', async () => {
  const app = global.createTestApp();

  app.post('/', body({ max: 9 }), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('is büffer: ' + Buffer.isBuffer(req.body) + ' ' + req.body.length);
    res.end();
  });

  await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(413)
    .send('0123456789');
});


src/__tests__/form.ts
import request from 'supertest';
import { form } from '..';

it('should parse request body to a key/value pair object', async () => {
  const app = global.createTestApp();

  app.post('/', form(), async (req, res) => {
    res.write(`typeof: ${typeof req.body} b=${req.body.b}; a=${req.body.a}`);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send('a=5979&b=%C3%84%C3%B6%C3%9C');

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('typeof: object b=ÄöÜ; a=5979');
});

it('should parse request body to a key/value pair and return 200 if max = 13 and body = 12', async () => {
  const app = global.createTestApp();

  app.post('/', form({ max: 13 }), async (req, res) => {
    res.write(`typeof: ${typeof req.body} b=${req.body.b}; a=${req.body.a}`);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send('a=5979&b=abc');

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('typeof: object b=abc; a=5979');
});

it('should parse request body to a key/value pair and return 200 if max = 12 and body = 12', async () => {
  const app = global.createTestApp();

  app.post('/', form({ max: 12 }), async (req, res) => {
    res.write(`typeof: ${typeof req.body} b=${req.body.b}; a=${req.body.a}`);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send('a=5979&b=abc');

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('typeof: object b=abc; a=5979');
});

it('should parse request body to a key/value pair and return 413 if body is bigger than max', async () => {
  const app = global.createTestApp();

  app.post('/', form({ max: 13 }), async (req, res) => {
    res.write('typeof: ' + (typeof req.body) + ' ' + req.body.length);
    res.end();
  });

  await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(413)
    .send('a=5979&b=abcde');
});


src/__tests__/json.ts
import request from 'supertest';
import { json, ParseErrorHandler } from '..';

it('should parse request body to a JSON object', async () => {
  const app = global.createTestApp();

  app.post('/', json(), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('typeöf: ' + (typeof req.body) + ' ' + req.body.length);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send({ length: 666 });

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('typeöf: object 666');
});

it('should parse request body to a JSON object and return 200 if max = 15 and body = 14', async () => {
  const app = global.createTestApp();

  app.post('/', json({ max: 15 }), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('typeöf: ' + (typeof req.body) + ' ' + req.body.length);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send(JSON.stringify({ length: 777 }));

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('typeöf: object 777');
});

it('should parse request body to a JSON object and return 200 if max = 15 and body = 15', async () => {
  const app = global.createTestApp();

  app.post('/', json({ max: 15 }), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('typeöf: ' + (typeof req.body) + ' ' + req.body.length);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(200)
    .send(JSON.stringify({ length: 7777 }));

  expect(Buffer.isBuffer(response.body)).toBe(true);
  expect(response.body.toString('utf8')).toBe('typeöf: object 7777');
});

it('should parse request body to a JSON object and return 413 if body is bigger than max', async () => {
  const app = global.createTestApp();

  app.post('/', json({ max: 13 }), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('typeöf: ' + (typeof req.body) + ' ' + req.body.length);
    res.end();
  });

  await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(413)
    .send(JSON.stringify({ length: 888 }));
});

it('should return 400 on invalid JSON', async () => {
  const app = global.createTestApp();

  app.post('/', json(), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('typeöf: ' + (typeof req.body) + ' ' + req.body.length);
    res.end();
  });

  await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(400)
    .send('{ length: 999 ');
});

it('should return 406 by custom parse error handler on invalid JSON', async () => {
  const app = global.createTestApp();

  const onParseFailed: ParseErrorHandler = async (ctx) => {
    ctx.response.writeHead(406);
    ctx.response.end(JSON.stringify(
      ctx.error.innerError.name + ': ' + ctx.error.innerError.message
    ));
  };

  app.post('/', json({ onParseFailed }), async (req, res) => {
    //            👇 this is no typo 😉
    res.write('typeöf: ' + (typeof req.body) + ' ' + req.body.length);
    res.end();
  });

  const response = await request(app)
    .post('/')
    .parse(global.parseBody)
    .expect(406)
    .send('{ length: 999 ');

  const message = JSON.parse(
    response.body.toString('utf8')
  );

  expect(typeof message).toBe('string');
  expect(message).toBe('SyntaxError: Unexpected token l in JSON at position 2');
});


index.d.ts
// Copyright 2020-present Marcel Joachim Kloubert <marcel.kloubert@gmx.net>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import flitz from 'flitz';

declare module 'flitz' {
  export interface Request {
    /**
     * The body, if parsed.
     */
    body?: any;
  }
}

src/index.ts
// Copyright 2020-present Marcel Joachim Kloubert <marcel.kloubert@gmx.net>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

/// <reference path="../index.d.ts" />

import { body } from './parsers';

exports = module.exports = function () {
  return body.apply(null, arguments as any);
};

export * from './errors';
export * from './parsers';
export * from './streams';

export default body;


src/errors.ts
// Copyright 2020-present Marcel Joachim Kloubert <marcel.kloubert@gmx.net>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

/**
 * Describes an 'entity too large' error.
 */
export class EntityTooLargeError extends Error {
}

/**
 * A parse error.
 */
export class ParseError extends Error {
  /**
   * Initializes a new instance of that class.
   *
   * @param {any} innerError The inner error.
   */
  public constructor(public readonly innerError: any) {
    super();
  }
}


src/streams.ts
// Copyright 2020-present Marcel Joachim Kloubert <marcel.kloubert@gmx.net>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import { CanBeNil } from 'flitz';
import { EntityTooLargeError } from './errors';

/**
 * Options for 'readStream()' function.
 */
export interface ReadStreamOptions {
  /**
   * Defines the maximum size for the stream, in bytes.
   */
  max?: CanBeNil<number>;
}

/**
 * Reads all data from a stream.
 *
 * @param {NodeJS.ReadableStream} stream The source stream.
 * @param {CanBeNil<ReadStreamOptions>} [options] Custom options.
 * 
 * @returns {Promise<Buffer>} The promise with the read data.
 */
export function readStream(stream: NodeJS.ReadableStream, options?: CanBeNil<ReadStreamOptions>): Promise<Buffer> {
  const max = options?.max;
  if (max) {
    if (typeof max !== 'number') {
      throw new TypeError('options.max must be a number');
    }

    if (max < 0) {
      throw new TypeError('options.max must be a greater than or equal 0');
    }
  }

  let concatData: (data: Buffer, chunk: Buffer) => Buffer;
  if (max) {
    // with max length
    concatData = (d, c) => {
      if (d.length + c.length > max) {
        throw new EntityTooLargeError();
      }

      return Buffer.concat([d, c]);
    };
  } else {
    concatData = (d, c) => Buffer.concat([d, c]);
  }

  return new Promise((resolve, reject) => {
    stream.once('error', err => {
      reject(err);
    });

    let data = Buffer.alloc(0);

    stream.on('data', (chunk: Buffer) => {
      try {
        data = concatData(data, chunk);
      } catch (e) {
        reject(e);
      }
    });

    stream.once('end', () => {
      resolve(data);
    });
  });
}

src/setupTests.ts
import flitz, { Flitz } from 'flitz';
import { Response } from 'supertest';

type Parser = (res: Response, callback: (err: Error | null, body: any) => void) => void;

declare global {
  namespace NodeJS {
    interface Global {
      /**
       * That function is ONLY available in test files!
       */
      createTestApp(): Flitz;
      /**
       * That function is ONLY available in test files!
       */
      parseBody: Parser;
    }
  }
}

global.createTestApp = () => {
  const app = flitz();

  return app;
};

global.parseBody = (res, callback) => {
  let data = Buffer.alloc(0);

  res.once('error', err => {
    callback(err, null);
  });

  res.on('data', (chunk: Buffer) => {
    data = Buffer.concat([data, chunk]);
  });

  res.once('end', () => {
    callback(null, data);
  });
};


src/parsers.ts
// Copyright 2020-present Marcel Joachim Kloubert <marcel.kloubert@gmx.net>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import querystring from 'querystring';
import { CanBeNil, Middleware, NextFunction, Request, RequestHandler, Response } from 'flitz';
import { safeLoad as loadYaml, YAMLException } from 'js-yaml';
import { EntityTooLargeError, ParseError } from './errors';
import { readStream } from './streams';

/**
 * Options for 'body()' function.
 */
export interface BodyOptions {
  /**
   * Defines the maximum size of the body, in bytes.
   */
  max?: CanBeNil<number>;
  /**
   * A custom handler, to tell the client, that the body is too big.
   */
  onMaxReached?: CanBeNil<RequestHandler>;
}

/**
 * Body options with parse error handling.
 */
export interface BodyOptionsWithParseErrorHandling extends BodyOptions {
  /**
   * A custom parse error handler.
   */
  onParseFailed?: CanBeNil<ParseErrorHandler>;
}

/**
 * Options for 'form()' function.
 */
export interface FormOptions extends BodyOptions {
}

/**
 * Options for 'json()' function.
 */
export interface JsonOptions extends BodyOptionsWithParseErrorHandling {
}

/**
 * Handles a body parse error.
 * 
 * @param {ParseErrorHandlerContext} context The context.
 */
export type ParseErrorHandler = (context: ParseErrorHandlerContext) => Promise<any>;

/**
 * A context of a parse error handler.
 */
export interface ParseErrorHandlerContext {
  /**
   * The error.
   */
  error: ParseError;
  /**
   * The request context.
   */
  request: Request;
  /**
   * The response context.
   */
  response: Response;
}

/**
 * Options for 'string()' function.
 */
export interface StringOptions extends BodyOptions {
}

/**
 * Options for 'yaml()' function.
 */
export interface YamlOptions extends BodyOptionsWithParseErrorHandling {
}

/**
 * Creates a new middleware that reads the complete
 * request body and writes the data to 'body' property
 * of request context.
 * 
 * @params {CanBeNil<BodyOptions>} [options] Custom options.
 * 
 * @returns {Middleware} The new middleware.
 */
export function body(options?: CanBeNil<BodyOptions>): Middleware {
  return withEntityTooLarge(async (req, res, next) => {
    req.body = await readStream(req, {
      max: options?.max
    });

    next();
  }, options?.onMaxReached);
}

/**
 * Creates a new middleware that reads the complete
 * request body and writes the data as key / value pairs
 * to 'body' property of request context.
 * 
 * @params {CanBeNil<FormOptions>} [options] Custom options.
 * 
 * @returns {Middleware} The new middleware.
 */
export function form(options?: CanBeNil<FormOptions>): Middleware {
  return withEntityTooLarge(async (req, res, next) => {
    req.body = querystring.parse(
      (await readStream(req, {
        max: options?.max
      })).toString('ascii')
    );

    next();
  }, options?.onMaxReached);
}

function getParseErrorHandler(options: CanBeNil<BodyOptionsWithParseErrorHandling>): ParseErrorHandler {
  let onParseFailed = options?.onParseFailed;
  if (!onParseFailed) {
    onParseFailed = async (ctx) => {
      if (!ctx.response.headersSent) {
        ctx.response.writeHead(400);
      }

      ctx.response.end();
    };
  }

  return onParseFailed;
}

/**
 * Creates a new middleware that reads the complete
 * request body and writes the data as JSON object
 * to 'body' property of request context.
 * 
 * @params {CanBeNil<JsonOptions>} [options] Custom options.
 * 
 * @returns {Middleware} The new middleware.
 */
export function json(options?: CanBeNil<JsonOptions>): Middleware {
  const onParseFailed = getParseErrorHandler(options);

  return withEntityTooLarge(async (request, response, next) => {
    const data = await readStream(request, {
      max: options?.max
    });

    try {
      request.body = data.length ? JSON.parse(data.toString('utf8')) : null;

      next();
    } catch (e) {
      if (e instanceof SyntaxError) {
        await onParseFailed({ error: new ParseError(e), request, response });
      } else {
        throw e;
      }
    }
  }, options?.onMaxReached);
}

/**
 * Creates a new middleware that reads the complete
 * request body and writes the data as string
 * to 'body' property of request context.
 * 
 * @params {CanBeNil<StringOptions>} [options] Custom options.
 * 
 * @returns {Middleware} The new middleware.
 */
export function string(options?: CanBeNil<StringOptions>): Middleware {
  return withEntityTooLarge(async (req, res, next) => {
    req.body = (await readStream(req, {
      max: options?.max
    })).toString('utf8');

    next();
  }, options?.onMaxReached);
}

/**
 * Creates a new middleware that reads the complete
 * request body as YAML document and writes the data as JSON object
 * to 'body' property of request context.
 * 
 * @params {CanBeNil<YamlOptions>} [options] Custom options.
 * 
 * @returns {Middleware} The new middleware.
 */
export function yaml(options?: CanBeNil<YamlOptions>): Middleware {
  const onParseFailed = getParseErrorHandler(options);

  return withEntityTooLarge(async (request, response, next) => {
    const data = await readStream(request, {
      max: options?.max
    });

    try {
      request.body = data.length ? loadYaml(data.toString('utf8')) : null;

      next();
    } catch (e) {
      if (e instanceof YAMLException) {
        await onParseFailed({ error: new ParseError(e), request, response });
      } else {
        throw e;
      }
    }
  }, options?.onMaxReached);
}

function withEntityTooLarge(
  action: (request: Request, response: Response, next: NextFunction) => Promise<void>,
  onMaxReached: CanBeNil<RequestHandler>
): Middleware {
  if (!onMaxReached) {
    // default handler
    onMaxReached = async (req, res) => {
      if (!res.headersSent) {
        res.writeHead(413);
      }

      res.end();
    };
  }

  return async (req, res, next) => {
    try {
      await action(req, res, next);
    } catch (err) {
      if (err instanceof EntityTooLargeError) {
        await onMaxReached!(req, res);
      } else {
        throw err;
      }
    }
  };
}